python/neuroglancer/ingest/tasks.py

```diff
@@ -1,66 +1,348 @@
+
+    def _generate_chunks(self, scale, chunk_size):
+        highest_resolution = np.array(self._info['scales'][0]['resolution'])
+        current_resolution = np.array(scale["resolution"])
+        downsample_ratio = current_resolution / highest_resolution
+        current_shape = self._data.shape[:-1] / downsample_ratio # we discard the channel component in data
+
+        # This is required because last data chunk is allowed to not 
+        # be a multiple of the chunk size.
+        n_chunks = self._ceil(current_shape, chunk_size)
+        for x,y,z in itertools.product(*list(map(xrange, n_chunks))):
+            # numpy allows for index that are larger than the array size
+            # because we are doing ceil in n_chunks, there will be cases where
+            # chunk.shape is less than chunk_size. In that case the filename should 
+            # base on chunk.shape.
+            chunk = self._data[
+                x * chunk_size[0] * downsample_ratio[0] : (x+1) * chunk_size[0] * downsample_ratio[0] : downsample_ratio[0],
+                y * chunk_size[1] * downsample_ratio[1] : (y+1) * chunk_size[1] * downsample_ratio[1] : downsample_ratio[1],
+                z * chunk_size[2] * downsample_ratio[2] : (z+1) * chunk_size[2] * downsample_ratio[2] : downsample_ratio[2],
+                :]
+
+            encoded = self._encode(chunk, scale["encoding"])
+            filename = self._get_filename(x, y, z, chunk_size, downsample_ratio, scale)
+            yield encoded, filename
+
+    def _encode(self, chunk, encoding):
+        if encoding == "jpeg":
+            return chunks.encode_jpeg(chunk)
+        elif encoding == "npz":
+            return chunks.encode_npz(chunk)
+        elif encoding == "raw":
+            return chunks.encode_raw(chunk)
+        else:
+            raise NotImplementedError(encoding)
+
+    def _ceil(self, arr1, arr2):
+        return np.ceil(arr1.astype(np.float32) / arr2).astype(np.uint32)
+
+    def _get_filename(self, x, y, z, chunk_size, downsample_ratio, scale):
+        xmin = x * chunk_size[0] + self._xmin / downsample_ratio[0]
+        xmax = min((x + 1) * chunk_size[0] + self._xmin / downsample_ratio[0], scale['size'][0])
+        ymin = y * chunk_size[1] + self._ymin / downsample_ratio[1]
+        ymax = min((y + 1) * chunk_size[1] + self._ymin / downsample_ratio[1], scale['size'][1])
+        zmin = z * chunk_size[2] + self._zmin / downsample_ratio[2]
+        zmax = min((z + 1) * chunk_size[2] + self._zmin / downsample_ratio[2], scale['size'][2])
+
+        return '{:d}-{:d}_{:d}-{:d}_{:d}-{:d}'.format(
+          xmin, xmax, ymin, ymax, zmin, zmax) 
+
+class DownsampleTask(object):
+    def __init__(self, chunk_path=None, info_path=None, fromjson=None, _id=None):
+        self._id =  None
+        self.chunk_path = chunk_path
+        self.info_path = info_path
+        self.tag = 'downsample'
+        if fromjson:
+            self.payloadBase64 = fromjson
```

Comment:
It actually uses the property setter. This line is equivalent to self.payloadBase64(fromjson)
Sorry, that's some tricky python syntax now that I think about. I guess that's the problem with this languages getter and setters they are sort of magical.
I'll never forget some code @chrisj wrote for thecube in which he did some like:
self.payloadBase64 = self.payloadBase64
which called the getter to update the setter.